types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+
+	struct wlr_drm_lease_device_v1 *device = lease->device;
+	wlr_drm_backend_terminate_lease(device->backend, lease->lessee_id);
+
+	for (size_t i = 0; i < lease->n_connectors; ++i) {
+		lease->connectors[i]->active_lease = NULL;
+	}
+}
+
+static void drm_lease_v1_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_drm_lease_v1 *lease = wlr_drm_lease_v1_from_resource(resource);
+	wlr_drm_lease_v1_revoke(lease);
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+	lease->resource = NULL;
+	free(lease->connectors);
+	free(lease);
+}
+
+static void drm_lease_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_v1_interface lease_impl = {
+	.destroy = drm_lease_v1_handle_destroy,
+};
+
+static void drm_lease_request_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_drm_lease_request_v1 *req =
+		wlr_drm_lease_request_v1_from_resource(resource);
+	wlr_log(WLR_DEBUG, "Destroying request %p", req);
+
+	free(req->connectors);
+	free(req);
+
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+}
+
+static void drm_lease_request_v1_handle_request_connector(
+		struct wl_client *client, struct wl_resource *request_resource,
+		struct wl_resource *connector_resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		wlr_drm_lease_request_v1_from_resource(request_resource);
+	struct wlr_drm_lease_connector_v1 *connector =
+		wlr_drm_lease_connector_v1_from_resource(connector_resource);
+
+	if (!connector || connector->active_lease) {
+		/* This connector offer has been withdrawn or is leased */
+		wlr_log(WLR_ERROR, "Failed to request connector");
+		request->invalid = true;
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Requesting connector %s", connector->output->name);
+
+	size_t n_connectors = request->n_connectors + 1;
+	request->connectors = realloc(request->connectors,
+			n_connectors * sizeof(struct wlr_drm_lease_connector_v1*));
+	request->connectors[request->n_connectors] = connector;
+	request->n_connectors = n_connectors;
+
+	/* Connector is not available for other clients */
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &connector->resources) {
+		if (resource != connector_resource) {
+			wp_drm_lease_connector_v1_send_withdrawn(resource);
+		}
+	}
+}
+
+static void drm_lease_request_v1_handle_submit(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_request_v1 *request =
+			wlr_drm_lease_request_v1_from_resource(resource);
+
+	struct wlr_drm_lease_v1 *lease = calloc(1, sizeof(struct wlr_drm_lease_v1));
+	if (!lease) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(client,
```

Comment:
The name `wl_resource` is confusing here, since there's also a `resource` variable. Use more detailed names like `lease_resource` and `request_resource`.